fdbserver/RestoreMaster.actor.h

```diff
@@ -90,98 +101,188 @@ struct RestoreMasterData : RestoreRoleData, public ReferenceCounted<RestoreMaste
-		// Assumption: fileIndex starts at 1. Each loader's initized fileIndex (NotifiedVersion type) starts at 0
-		int fileIndex = 0; // fileIndex must be unique; ideally it continuously increase across verstionBatches for
-		                   // easier progress tracking
-		int versionBatchId = 1;
-		for (auto versionBatch = versionBatches->begin(); versionBatch != versionBatches->end(); versionBatch++) {
-			std::sort(versionBatch->second.rangeFiles.begin(), versionBatch->second.rangeFiles.end());
-			std::sort(versionBatch->second.logFiles.begin(), versionBatch->second.logFiles.end());
-			for (auto& logFile : versionBatch->second.logFiles) {
-				logFile.fileIndex = ++fileIndex;
-				TraceEvent("FastRestore")
-				    .detail("VersionBatchId", versionBatchId)
-				    .detail("LogFile", logFile.toString());
-			}
-			for (auto& rangeFile : versionBatch->second.rangeFiles) {
-				rangeFile.fileIndex = ++fileIndex;
-				TraceEvent("FastRestore")
-				    .detail("VersionBatchId", versionBatchId)
-				    .detail("RangeFile", rangeFile.toString());
+		int logIdx = 0;
+		std::vector<RestoreFileFR> retLogs;
+		// Scan all logFiles every time to avoid assumption on log files' version ranges.
+		// For example, we do not assume each version range only exists in one log file
+		while (logIdx < logFiles.size()) {
+			Version begin = std::max(prevVersion, logFiles[logIdx].beginVersion);
+			Version end = std::min(nextVersion, logFiles[logIdx].endVersion);
+			if (begin < end) { // logIdx file overlap in [prevVersion, nextVersion)
+				double ratio = (end - begin) / (logFiles[logIdx].endVersion - logFiles[logIdx].beginVersion);
+				size += logFiles[logIdx].fileSize * ratio;
+				retLogs.push_back(logFiles[logIdx]);
 			}
-			versionBatchId++;
+			++logIdx;
 		}
+		return std::make_tuple(size, rangeIdx, retLogs);
+	}
 
-		TraceEvent("FastRestore").detail("VersionBatches", versionBatches->size());
-		// Sanity check
-		std::set<uint32_t> fIndexSet;
-		for (auto& versionBatch : *versionBatches) {
-			Version prevVersion = 0;
-			for (auto& logFile : versionBatch.second.logFiles) {
-				TraceEvent("FastRestore").detail("PrevVersion", prevVersion).detail("LogFile", logFile.toString());
-				ASSERT(logFile.beginVersion >= versionBatch.second.beginVersion);
-				ASSERT(logFile.endVersion <= versionBatch.second.endVersion);
-				ASSERT(prevVersion <= logFile.beginVersion);
-				prevVersion = logFile.endVersion;
-				ASSERT(fIndexSet.find(logFile.fileIndex) == fIndexSet.end());
-				fIndexSet.insert(logFile.fileIndex);
+	// Split backup files into version batches, each of which has similar data size
+	// Input: sorted range files, sorted log files;
+	// Output: a set of version batches whose size is less than opConfig.batchSizeThreshold
+	//    	   and each mutation in backup files is included in the version batches exactly once.
+	// Assumption 1: input files has no empty files;
+	// Assumption 2: range files at one version <= batchSizeThreshold.
```

Comment:
Another option is to allow a batch to include one version that causes it to exceed the threshold. That way you don't have failures when this assumption doesn't hold. 

You could either use this to simplify the logic a bit (i.e. you don't need the rewrite next version check), or you could push any version that exceeds the size limit to the next batch and then allow it to go through if it still exceeds the limit.

Reply:
Because the data will be buffered in the cluster until all data in the version batch has been committed to the destination cluster. 
If it exceeds the limit, it may cause out of memory problem on some host. 
That's why the extra complexity is involved in this function. 

If we want to host more data in a version batch, DBA can simply set a larger threshold.

Reply:
Yeah, I guess the important question is how big can a version be? We have precedent for using the logic that I described where the limit is a soft limit allowing only a single version to exceed it. This makes dealing with the limit a little bit easier and also avoids the problem where things break when a version gets too big. If the size of a single version is bounded and not unreasonable, I think I would prefer that approach so that no special configuration is necessary. If it isn't bounded, then I agree we can't do it.